### ISSUE TITLE
OpenAI Assistants - continue conversation

### ISSUE THREAD

**Issue by jackitaliano** [OvidijusParsiunas/deep-chat#115]
# Scenario
- `directConnection` to OpenAI assistants
- attempting to save chat history and multiple chats via threads

# My thought for solution
1. Deepchat consumer saves OpenAI thread id in local storage
2. Deepchat allows setting `thread_id` in `onComponentRender` similar to what can be done for `chatElementRef.directConnection.openAI.assistant.assistant_id`
3. Either deepchat loads conversation history from thread, or consumer loads messages themselves with `initialMessages`.
  a. If consumer loads message, deepchat would still retain `thread_id` set by consumer

# My request
- Allow the consumer to set openai assistant's active thread_id in `onComponentRender`

# Reason
- OpenAI's API doesn't allow adding 'Assistant' messages
  - consumer can use `initialMessages` or `addMessage` to insert messages into deepchat, but this doesn't give the assistant the messages. The messages will display, but the assistant has no knowledge of them.
  
 # Use Case
- Allows consumer to handle saving and loading of state for OpenAI assistants
  - Allows for saving of chat history
  - Allows for multiple chats

# Similar Requests
[Save and Retrieve Chat Sessions | Issue #49](https://github.com/OvidijusParsiunas/deep-chat/issues/49)

In the comments of this request, you mentioned adding a way to *load messages*, but merely loading messages in the UI is not as useful as loading a thread for the purposes of an assistant.

# Example Psuedocode
```javascript
<deep-chat
    onComponentRender = function() {
        ...
        const local_thread_id = localstorage.getItem('openai-thread-id');
        if (local_thread_id) {
            chatElementRef.directConnection.openAI.assistant.thread_id = local_thread_id;
        }
        ...
    }
...
/>
```

I'm open to other solutions to this issue if you have any ideas, or maybe I'm missing something that might already solve this (though I feel I have thoroughly searched).

The work you're doing is fantastic, thank you very much!

**Comment by OvidijusParsiunas** [comment]
Hi @jackitaliano. Just so I understand correctly, you want to set the thread_id and automatically start the chat with the previous messages being pre-populated in the chat box?

**Comment by jackitaliano** [comment]
> Hi @jackitaliano. Just so I understand correctly, you want to set the thread_id and automatically start the chat with the previous messages being pre-populated in the chat box?

Hello @OvidijusParsiunas 

That would be the ideal solution for me. 

The main thing is being able to set the thread_id so the assistant has the context of thread. With the current deep-chat api and docs, it is certainly possible for me to pre-populate the previous messages in the chat box; however, I am not able to give those messages to the Assistant as I'd like to.

In the case where no thread_id is set by the consumer, it would default to it's current functionality.

If you're willing/able to implement both setting the thread_id *and* pre-populating the chat, that would be even better.

Thanks for the response!

**Comment by jackitaliano** [comment]
> Let me see what I can do. I'll try out a couple of things and give you an update on what's possible. Depending on the complexity of the solution I may need to continue this tomorrow as it is quite late where I live. I'll give you an update as soon as I have something. Thanks!

Great! No problem at all if you're not able to complete it tonight. Thanks.

**Comment by jackitaliano** [comment]
This works great, thank you very much! 

For anyone seeing this in the future, this is roughly my first shot implementation: (not exact because contains validation/logic not relevant here):

(svelte)
```javascript
async function getThread() { return localStorage.getItem('openai-thread-id'); }
async function setThread(thread_id) { localStorage.setItem('openai-thread-id', thread_id); }

let openAIThreadIdSet = false;

{#await getThread() then thread_id}
<deep-chat
    onNewMessage = {async function(message) {
        ...
        if (!openAIThreadIdSet && message.message._sessionId && message.message._sessionId != thread_id) {
          setThread(message.message._sessionId);
          openAIThreadIdSet = true;
        }
        ...
    }}
     directConnection={{
        openAI: {
          ...
          assistant: {
            ...
            thread_id: thread_id,
            load_thread_history: thread_id ? true : false,
            ...
          }
        }
      }}
...
/>
```

**Comment by OvidijusParsiunas** [comment]
This has now been released in Deep Chat version `1.4.11`. Checkout the updated [assistant](https://deepchat.dev/docs/directConnection/OpenAI#Assistant) documentation for more.
